### ISSUE TITLE
macos+homebrew gcc missing __has_feature()

### ISSUE THREAD

**Issue by egorpugin** [grpc/grpc#29680]
grpc 1.46.1
macos aarch64
gcc11 from homebrew
https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/port_platform.h#L241=

Build error - gcc has no `__has_feature` thing.

Error
```
[org.sw.demo.google.grpc.core.plugin_registry-1.46.1]/src/core/plugin_registry/grpc_plugin_registry_extra.cc
In file included from /Users/egor/.sw/storage/pkg/45/1d/cd62/src/sdir/include/grpc/support/port_platform.h:22,
                 from /Users/egor/.sw/storage/pkg/55/a8/fe29/src/sdir/src/core/plugin_registry/grpc_plugin_registry_extra.cc:17:
/Users/egor/.sw/storage/pkg/45/1d/cd62/src/sdir/include/grpc/impl/codegen/port_platform.h:241:46: error: missing binary operator before token "("
  241 | #if !(defined(__has_feature) && __has_feature(cxx_thread_local))
      |                                              ^
```

**Comment by egorpugin** [comment]
iOS ifdef ends on line 240. Line 241 is for common `__APPLE__`.

**Comment by egorpugin** [comment]
Agreed. Motivation is understandable.

**Comment by barracuda156** [comment]
Not a Homebrew issue, building with `gcc12` fails too on the same: https://trac.macports.org/ticket/65525

**Comment by barracuda156** [comment]
Still unfixed in 1.48.4.

**Comment by barracuda156** [comment]
@nicolasnoble It is an actual bug, the code should be:
```
#if defined(__has_feature)
#if __has_feature(cxx_thread_local)
#define GPR_PTHREAD_TLS 1
#endif
#endif
```
See: https://trac.macports.org/ticket/65525#comment:1

**Comment by ryandesign** [comment]
> the code should be:
> 
> ```
> #if defined(__has_feature)
> #if __has_feature(cxx_thread_local)
> #define GPR_PTHREAD_TLS 1
> #endif
> #endif
> ```

No, that's backwards. This is the code you were proposing to replace:

https://github.com/grpc/grpc/blob/b37996d50e3b2f55875ea5a701da2969e7f1870a/include/grpc/impl/codegen/port_platform.h#L237-L239

It defines `GPR_PTHREAD_TLS` if the C++ compiler does *not* support thread-local storage, in which case it means that gprc should provide a compatibility implementation of TLS that uses pthread.

The proposed new code defines `GPR_PTHREAD_TLS` if the C++ compiler *does* support TLS, meaning that if the compiler supports TLS, gprc will not use that support, and for old compilers that don't support TLS, the code will presumably not compile.

The code should have been:

```c
#ifndef __has_feature
#define GPR_PTHREAD_TLS 1
#elif !__has_feature(cxx_thread_local)
#define GPR_PTHREAD_TLS 1
#endif
```

There's already a `GPR_HAS_FEATURE` macro:

https://github.com/grpc/grpc/blob/b37996d50e3b2f55875ea5a701da2969e7f1870a/include/grpc/impl/codegen/port_platform.h#L663-L669

so it should be possible to simplify it to:

```c
#if !GPR_HAS_FEATURE(cxx_thread_local)
#define GPR_PTHREAD_TLS 1
#endif
```

except for the complication that `GPR_HAS_FEATURE` isn't defined until later in the file.

The whole pthread-based compatibility implementation of thread-local storage was removed in f15ba1ffc73c5d0b63754df72fc47c567b9c7252 in gprc 1.51.0 so this issue can be closed.

**Comment by barracuda156** [comment]
@ryandesign I just picked Josh’s suggestion, given the fact that exiting code was wrong and had to be fixed.
